### ISSUE TITLE
decode error caused by range query

### ISSUE THREAD

**Issue by ianhe8x** [polkadot-js/api#3104]
Query against polkadot archive node
```
await api.query.system.events.range(['0xb70008a998e07bac35474412f0a1dff12b0460bb3d5c834ce1db915bb8a38f6a', '0x80849c3c4c87820a60e7b278819821abbcdccf61a55ca8da0986137429ade760'])
```

get this error
```
Uncaught Error: createType(Vec<EventRecord>):: Vec length 2759856644 exceeds 65536
```

RPC logs:
```
2021-01-25 17:44:05          API-WS: calling state_queryStorage {"id":154,"jsonrpc":"2.0","method":"state_queryStorage","params":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7"],"0xb70008a998e07bac35474412f0a1dff12b0460bb3d5c834ce1db915bb8a38f6a","0x80849c3c4c87820a60e7b278819821abbcdccf61a55ca8da0986137429ade760"]}
2021-01-25 17:44:05          API-WS: received {"jsonrpc":"2.0","result":[{"block":"0xb70008a998e07bac35474412f0a1dff12b0460bb3d5c834ce1db915bb8a38f6a","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x0c0000000000000080e36a09000000000200000001000000000000000000000000000200000002000000000000ca9a3b00000000020000"]]},{"block":"0x90d68ed01bec444ca8a487874a13942426b42ee6d70823e2f6d2d3a90616a385","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x180000000000000080e36a09000000000200000001000000000000000000000000000200000002000000000000ca9a3b0000000002000000030000000e0600127a0000000000000000000000000000000300000003041280a479ee3beca7af1636aca17582f30829782e2c9b1b9c72aaf8060563ab3780841e00000000000000000000000000000003000000000040c00b1200000000000000"]]},{"block":"0xb93f531b5703229a2a7c7d31c82db4f234d4ea9fed774b7ad80ba2d20355ae13","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x080000000000000080e36a09000000000200000001000000000000ca9a3b00000000020000"]]},{"block":"0x2dcd984c10b8e1793b06412834312ec504fe8078524cc35ec112fdcf297cb574","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x0c0000000000000080e36a09000000000200000001000000000000000000000000000200000002000000000000ca9a3b00000000020000"]]},{"block":"0x9a75a5cdcb7a20c5981084ffdca13f430a68e066267f3a98275a1beff332fa48","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x080000000000000080e36a09000000000200000001000000000000ca9a3b00000000020000"]]},{"block":"0x8ab9fc7afffdb41486772a948bf56f4a8cd7f73ed3eccc1b0f136fd3d82473e6","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x0c0000000000000080e36a09000000000200000001000000000000000000000000000200000002000000000000ca9a3b00000000020000"]]},{"block":"0x4e16ed2e068e9fd86c703bb76b35dea9889299748c78916fd5f171b1b33a5dfe","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x080000000000000080e36a09000000000200000001000000000000ca9a3b00000000020000"]]},{"block":"0xa43717ab3511b0abc4dc228399f551b2afe17ba107d44499363a70d355f8f6a7","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x0c0000000000000080e36a09000000000200000001000000000000000000000000000200000002000000000000ca9a3b00000000020000"]]},{"block":"0x09c629176cf75ee7a8b1fc4146f4406f9ecda68b94b5c23723214ee2c6895c9f","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x080000000000000080e36a09000000000200000001000000000000ca9a3b00000000020000"]]},{"block":"0xfa41ab9f343dcdab398d2d48056c0f6539555d59f6dbbd7990afad0bee7a3a51","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x0c0000000000000080e36a09000000000200000001000000000000000000000000000200000002000000000000ca9a3b00000000020000"]]},{"block":"0x246c4aabbedc37c5278a2316b2c1f95c0bd7a06f1da7a629d2b959dd0d2768a6","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x080000000000000080e36a09000000000200000001000000000000ca9a3b00000000020000"]]},{"block":"0x8d14a1dae6478400046f75a18a4689746ac1aa9713aef7846cde2c9ad0dd583d","changes":[["0x26aa394eea5630e07c48ae0c9558cef780d41e5e16056765bc8461851072c9d7","0x0c0000000000000080e36a09000000000200000001000000000000000000000000000200000002000000000000ca9a3b00000000020000"]]}],"id":154}

```

**Comment by jacogr** [comment]
There is an issue with range - unlike for instance with `.at` it does not auto-decorate (and cache) the relevant metadata for specific blocks. Before decoding it would need to retrieve the version for each hash retrieved, then attach the metadata if not yet available/cached and only then decode. 

This is known (but not documented), has not bubbled up the list of priorities at all since exposing the relevant safe RPCs are generally not a good idea on public nodes. (Obviously it does have a nice niche it can fill)

The hoops to get this to work is not entirely simple, but at least the process has been done for `.at` queries - just would need some thinking about how to best do it in the actual `.range` decoding loops.

Basically, this is what is done for `.at` queries - https://github.com/polkadot-js/api/blob/master/packages/rpc-core/src/index.ts#L242-L255

And here is where the `.range` gets decoded - https://github.com/polkadot-js/api/blob/master/packages/rpc-core/src/index.ts#L395-L405

**Comment by jacogr** [comment]
Closing. Won't be fixed - range is marked deprecated, will be removed. (Not easily fixable due to having to check each block for versions, this is not really something we want to maintain here)
